File: solutions/128LongestConsecutiveSequence/main.go
/*
Given an unsorted array of integers nums, return the
length of the longest consecutive elements sequence.

You must write an algorithm that runs in O(n) time.
*/

func longestConsecutive(nums []int) int {
	set := map[int]bool{}
	for _, num := range nums {
		set[num] = true
	}
	maxSequence := 0
	for _, num := range nums {
		if set[num-1] {
			continue
		}
		sequence, tmp := 1, num+1
		for set[tmp] {
			sequence++
			tmp++
		}
		maxSequence = max(maxSequence, sequence)
		if sequence > len(nums)/2 {
			break
		}
	}
	return maxSequence
}

File: solutions/128LongestConsecutiveSequence/main.py
'''
Given an unsorted array of integers nums, return the
length of the longest consecutive elements sequence.

You must write an algorithm that runs in O(n) time.
'''

class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        nums_set = set(nums)
        max_sequence = 0
        for num in nums:
            if num-1 in nums_set:
                continue
            sequence, tmp = 1, num+1
            while tmp in nums_set:
                sequence += 1
                tmp += 1
            max_sequence = max(max_sequence, sequence)
            if sequence > len(nums) / 2:
                break
        return max_sequence